File: MVS/Development/Samples/C#/MvCameraControlNet_V2/AreaScanCamera/ParametrizeCamera_AreaScanIOSettings/ParametrizeCamera_AreaScanIOSettings.cs
/*
 * 这个示例演示了从相机中获取配置文件。
 * This sample shows how to obtain configuration files from a camera.
 */

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using MvCameraControl;
using System.Threading;

namespace ParameterCamera_AreaScanIOSettings
{
    class ParameterCamera_AreaScanIOSettings
    {
        private const DeviceTLayerType devLayerType = DeviceTLayerType.MvGigEDevice | DeviceTLayerType.MvUsbDevice | DeviceTLayerType.MvGenTLCameraLinkDevice
           | DeviceTLayerType.MvGenTLCXPDevice | DeviceTLayerType.MvGenTLXoFDevice;


        void FrameGrabedEventHandler(object sender, FrameGrabbedEventArgs e)
        {
            Console.WriteLine("Get one frame: Width[{0}] , Height[{1}] , ImageSize[{2}], FrameNum[{3}]", e.FrameOut.Image.Width, e.FrameOut.Image.Height, e.FrameOut.Image.ImageSize, e.FrameOut.FrameNum);
        }

        public void Run()
        {
            IDevice device = null;

            try
            {
                List<IDeviceInfo> devInfoList;

                // ch:枚举设备 | en:Enum device
                int ret = DeviceEnumerator.EnumDevices(devLayerType, out devInfoList);
                if (ret != MvError.MV_OK)
                {
                    Console.WriteLine("Enum device failed:{0:x8}", ret);
                    return;
                }

                Console.WriteLine("Enum device count : {0}", devInfoList.Count);

                if (0 == devInfoList.Count)
                {
                    return;
                }

                // ch:打印设备信息 en:Print device info
                int devIndex = 0;
                foreach (var devInfo in devInfoList)
                {
                    Console.WriteLine("[Device {0}]:", devIndex);
                    if (devInfo.TLayerType == DeviceTLayerType.MvGigEDevice || devInfo.TLayerType == DeviceTLayerType.MvVirGigEDevice || devInfo.TLayerType == DeviceTLayerType.MvGenTLGigEDevice)
                    {
                        IGigEDeviceInfo gigeDevInfo = devInfo as IGigEDeviceInfo;
                        uint nIp1 = ((gigeDevInfo.CurrentIp & 0xff000000) >> 24);
                        uint nIp2 = ((gigeDevInfo.CurrentIp & 0x00ff0000) >> 16);
                        uint nIp3 = ((gigeDevInfo.CurrentIp & 0x0000ff00) >> 8);
                        uint nIp4 = (gigeDevInfo.CurrentIp & 0x000000ff);
                        Console.WriteLine("DevIP: {0}.{1}.{2}.{3}", nIp1, nIp2, nIp3, nIp4);
                    }

                    Console.WriteLine("ModelName:" + devInfo.ModelName);
                    Console.WriteLine("SerialNumber:" + devInfo.SerialNumber);
                    Console.WriteLine();
                    devIndex++;
                }

                Console.Write("Please input index(0-{0:d}):", devInfoList.Count - 1);

                devIndex = Convert.ToInt32(Console.ReadLine());

                if (devIndex > devInfoList.Count - 1 || devIndex < 0)
                {
                    Console.Write("Input Error!\n");
                    return;
                }

                // ch:创建设备 | en:Create device
                device = DeviceFactory.CreateDevice(devInfoList[devIndex]);

                // ch:打开设备 | en:Open device
                ret = device.Open();
                if (ret != MvError.MV_OK)
                {
                    Console.WriteLine("Open device failed:{0:x8}", ret);
                    return;
                }

                // ch:探测网络最佳包大小(只对GigE相机有效) | en:Detection network optimal package size(It only works for the GigE camera)
                if (device is IGigEDevice)
                {
                    int packetSize;
                    ret = (device as IGigEDevice).GetOptimalPacketSize(out packetSize);
                    if (packetSize > 0)
                    {
                        ret = device.Parameters.SetIntValue("GevSCPSPacketSize", packetSize);
                        if (ret != MvError.MV_OK)
                        {
                            Console.WriteLine("Warning: Set Packet Size failed {0:x8}", ret);
                        }
                        else
                        {
                            Console.WriteLine("Set PacketSize to {0}", packetSize);
                        }
                    }
                    else
                    {
                        Console.WriteLine("Warning: Get Packet Size failed {0:x8}", ret);
                    }
                }


                //Ch: 设置Line0 IO输入(也可选择Line2作为输入, 延迟更低) | en: Set Line0 to IO input(You can also choose Line2 as Input IO, which has Lower latency)
                Console.WriteLine("Set IO input...");
                {
                    // ch:设置触发模式为on | en:Set trigger mode as on
                    ret = device.Parameters.SetEnumValueByString("TriggerMode", "On");
                    if (ret != MvError.MV_OK)
                    {
                        Console.WriteLine("Set trigger mode fail! nRet {0:x8}", ret);
                        return;
                    }

                    // ch:设置触发源为Line0 | en:Set trigger source as Line0
                    ret = device.Parameters.SetEnumValueByString("TriggerSource", "Line0");
                    if (ret != MvError.MV_OK)
                    {
                        Console.WriteLine("Set trigger source fail! nRet {0:x8}", ret);
                        return;
                    }

                    // ch:设置触发沿为RisingEdge | en: Set TriggerActivation as RisingEdge
                    ret = device.Parameters.SetEnumValueByString("TriggerActivation", "RisingEdge");
                    if (ret != MvError.MV_OK)
                    {
                        Console.WriteLine("Set trigger activation RisingEdge fail! nRet {0:x8}", ret);
                        return;
                    }

                    // ch:设置触发延迟 | en: Set TriggerDelay
                    ret = device.Parameters.SetFloatValue("TriggerDelay", 0);
                    if (ret != MvError.MV_OK)
                    {
                        Console.WriteLine("Set trigger delay fail! nRet {0:x8}", ret);
                        return;
                    }

                    // ch:关闭触发缓存、如需要可开启 | en: Turn off TriggerCacheEnable 
                    ret = device.Parameters.SetBoolValue("TriggerCacheEnable", false);
                    if (ret != MvError.MV_OK)
                    {
                        Console.WriteLine("Set trigger cache enable fail! nRet {0:x8}", ret);
                        return;
                    }

                    // ch:切换LineSelector 为Line0 | en: Set LineSelector as Line0
                    ret = device.Parameters.SetEnumValueByString("LineSelector", "Line0");
                    if (ret != MvError.MV_OK)
                    {
                        Console.WriteLine("Set line selector fail! nRet {0:x8}", ret);
                        return;
                    }

                    // ch:设置Line0 滤波时间(us)、误触发可适当加大 | Set Line0 LineDebouncerTime(us)
                    ret = device.Parameters.SetIntValue("LineDebouncerTime", 50);
                    if (ret != MvError.MV_OK)
                    {
                        Console.WriteLine("Set line debouncer time fail! nRet {0:x8}\n", ret);
                        return;
                    }
                }

                //ch: 设置Line1 IO输出(也可选择Line2作为输出, 延迟更低), 用于控制外部光源、等设备 | en: Set Line1 as IO output (You can also choose Line2 as Output IO, which has Lower latency )
                Console.WriteLine("Set IO output...");
                {
                    // ch:切换LineSelector 为Line1 | en:Set LineSelector as Line1
                    ret = device.Parameters.SetEnumValueByString("LineSelector", "Line1");
                    if (ret != MvError.MV_OK)
                    {
                        Console.WriteLine("Set line selector fail! nRet {0:x8}", ret);
                        return;
                    }
                    //ch: 输出源选择曝光开始 |en:Set LineSource as ExposureStartActive
                    ret = device.Parameters.SetEnumValueByString("LineSource", "ExposureStartActive");
                    if (ret != MvError.MV_OK)
                    {
                        Console.WriteLine("Set LineSource fail! nRet {0:x8}", ret);
                        return;
                    }
                    // ch:开启输出使能| en: Turn on StrobeEnable
                    ret = device.Parameters.SetBoolValue("StrobeEnable", true);
                    if (ret != MvError.MV_OK)
                    {
                        Console.WriteLine("Set strobe enable fail! nRet {0:x8}", ret);
                        return;
                    }

                    // ch:设置输出线路持续时间（us） | en: Set StrobeLineDuration(us)
                    ret = device.Parameters.SetIntValue("StrobeLineDuration", 0);
                    if (ret != MvError.MV_OK)
                    {
                        Console.WriteLine("Set line strobe line duration fail! nRet {0:x8}", ret);
                        return;
                    }

                    // ch:设置输出线路持续时间（us） | en: Set StrobeLineDuration(us)
                    ret = device.Parameters.SetIntValue("StrobeLineDelay", 0);
                    if (ret != MvError.MV_OK)
                    {
                        Console.WriteLine("Set line strobe line delay fail! nRet {0:x8}", ret);
                        return;
                    }

                    // ch:输出线路预延迟（us） | en: Set StrobeLinePreDelay(us)
                    ret = device.Parameters.SetIntValue("StrobeLinePreDelay", 0);
                    if (ret != MvError.MV_OK)
                    {
                        Console.WriteLine("Set line strobe line pre-delay fail! nRet {0:x8}", ret);
                        return;
                    }
                }

                //ch: 设置合适的缓存节点数量 | en: Setting the appropriate number of image nodes
                device.StreamGrabber.SetImageNodeNum(5);

                // ch:注册回调函数 | en:Register image callback
                device.StreamGrabber.FrameGrabedEvent += FrameGrabedEventHandler;
                // ch:开启抓图 || en: start grab image
                ret = device.StreamGrabber.StartGrabbing();
                if (ret != MvError.MV_OK)
                {
                    Console.WriteLine("Start grabbing failed:{0:x8}", ret);
                    return;
                }

                Console.WriteLine("Press enter to exit");
                Console.ReadLine();


                // ch:停止抓图 | en:Stop grabbing
                ret = device.StreamGrabber.StopGrabbing();
                if (ret != MvError.MV_OK)
                {
                    Console.WriteLine("Stop grabbing failed:{0:x8}", ret);
                    return;
                }

                // ch:关闭设备 | en:Close device
                ret = device.Close();
                if (ret != MvError.MV_OK)
                {
                    Console.WriteLine("Close device failed:{0:x8}", ret);
                    return;
                }
            }
            catch (Exception e)
            {
                Console.Write("Exception: " + e.Message);
            }
            finally
            {
                // ch:销毁设备 | en:Destroy device
                if (device != null)
                {
                    device.Dispose();
                    device = null;
                }

            }
        }


        static void Main(string[] args)
        {
            // ch: 初始化 SDK | en: Initialize SDK
            SDKSystem.Initialize();

            ParameterCamera_AreaScanIOSettings program = new ParameterCamera_AreaScanIOSettings();
            program.Run();

            Console.WriteLine("Press enter to exit");
            Console.ReadKey();

            // ch: 反初始化SDK | en: Finalize SDK
            SDKSystem.Finalize();
        }
    }
}

File: MVS/Development/Samples/C++/AreaScanCamera/ParametrizeCamera_AreaScanIOSettings/ParametrizeCamera_AreaScanIOSettings.cpp
#include <stdio.h>
#include <Windows.h>
#include <process.h>
#include <conio.h>
#include "MvCameraControl.h"

bool g_bExit = false;

// ch:等待按键输入 | en:Wait for key press
void WaitForKeyPress(void)
{
    while(!_kbhit())
    {
        Sleep(10);
    }
    _getch();
}

bool PrintDeviceInfo(MV_CC_DEVICE_INFO* pstMVDevInfo)
{
    if (NULL == pstMVDevInfo)
    {
        printf("The Pointer of pstMVDevInfo is NULL!\n");
        return false;
    }
    if (pstMVDevInfo->nTLayerType == MV_GIGE_DEVICE)
    {
        int nIp1 = ((pstMVDevInfo->SpecialInfo.stGigEInfo.nCurrentIp & 0xff000000) >> 24);
        int nIp2 = ((pstMVDevInfo->SpecialInfo.stGigEInfo.nCurrentIp & 0x00ff0000) >> 16);
        int nIp3 = ((pstMVDevInfo->SpecialInfo.stGigEInfo.nCurrentIp & 0x0000ff00) >> 8);
        int nIp4 = (pstMVDevInfo->SpecialInfo.stGigEInfo.nCurrentIp & 0x000000ff);

        // ch:打印当前相机ip和用户自定义名字 | en:print current ip and user defined name
        printf("CurrentIp: %d.%d.%d.%d\n" , nIp1, nIp2, nIp3, nIp4);
        printf("UserDefinedName: %s\n\n" , pstMVDevInfo->SpecialInfo.stGigEInfo.chUserDefinedName);
    }
    else if (pstMVDevInfo->nTLayerType == MV_USB_DEVICE)
    {
        printf("UserDefinedName: %s\n", pstMVDevInfo->SpecialInfo.stUsb3VInfo.chUserDefinedName);
        printf("Serial Number: %s\n", pstMVDevInfo->SpecialInfo.stUsb3VInfo.chSerialNumber);
        printf("Device Number: %d\n\n", pstMVDevInfo->SpecialInfo.stUsb3VInfo.nDeviceNumber);
    }
    else if (pstMVDevInfo->nTLayerType == MV_GENTL_GIGE_DEVICE)
    {
        printf("UserDefinedName: %s\n", pstMVDevInfo->SpecialInfo.stGigEInfo.chUserDefinedName);
        printf("Serial Number: %s\n", pstMVDevInfo->SpecialInfo.stGigEInfo.chSerialNumber);
        printf("Model Name: %s\n\n", pstMVDevInfo->SpecialInfo.stGigEInfo.chModelName);
    }
    else if (pstMVDevInfo->nTLayerType == MV_GENTL_CAMERALINK_DEVICE)
    {
        printf("UserDefinedName: %s\n", pstMVDevInfo->SpecialInfo.stCMLInfo.chUserDefinedName);
        printf("Serial Number: %s\n", pstMVDevInfo->SpecialInfo.stCMLInfo.chSerialNumber);
        printf("Model Name: %s\n\n", pstMVDevInfo->SpecialInfo.stCMLInfo.chModelName);
    }
    else if (pstMVDevInfo->nTLayerType == MV_GENTL_CXP_DEVICE)
    {
        printf("UserDefinedName: %s\n", pstMVDevInfo->SpecialInfo.stCXPInfo.chUserDefinedName);
        printf("Serial Number: %s\n", pstMVDevInfo->SpecialInfo.stCXPInfo.chSerialNumber);
        printf("Model Name: %s\n\n", pstMVDevInfo->SpecialInfo.stCXPInfo.chModelName);
    }
    else if (pstMVDevInfo->nTLayerType == MV_GENTL_XOF_DEVICE)
    {
        printf("UserDefinedName: %s\n", pstMVDevInfo->SpecialInfo.stXoFInfo.chUserDefinedName);
        printf("Serial Number: %s\n", pstMVDevInfo->SpecialInfo.stXoFInfo.chSerialNumber);
        printf("Model Name: %s\n\n", pstMVDevInfo->SpecialInfo.stXoFInfo.chModelName);
    }
    else
    {
        printf("Not support.\n");
    }

    return true;
}

void __stdcall ImageCallBackEx(unsigned char * pData, MV_FRAME_OUT_INFO_EX* pFrameInfo, void* pUser)
{
	if (pFrameInfo)
	{
		printf("Get one Frame: width[%d], height[%d], nFrameNum[%d]\n", 
			pFrameInfo->nExtendWidth, pFrameInfo->nExtendHeight, pFrameInfo->nFrameNum);
	}
}

int main()
{
    int nRet = MV_OK;
    void* handle = NULL;

    do 
    {
		// ch:初始化SDK | en:Initialize SDK
		nRet = MV_CC_Initialize();
		if (MV_OK != nRet)
		{
			printf("Initialize SDK fail! nRet [0x%x]\n", nRet);
			break;
		}

        // ch:枚举设备 | en:Enum device
        MV_CC_DEVICE_INFO_LIST stDeviceList;
        memset(&stDeviceList, 0, sizeof(MV_CC_DEVICE_INFO_LIST));
        nRet = MV_CC_EnumDevices(MV_GIGE_DEVICE | MV_USB_DEVICE | MV_GENTL_CAMERALINK_DEVICE | MV_GENTL_CXP_DEVICE | MV_GENTL_XOF_DEVICE, &stDeviceList);
        if (MV_OK != nRet)
        {
            printf("Enum devices fail! nRet [0x%x]\n", nRet);
            break;
        }

        if (stDeviceList.nDeviceNum > 0)
        {
            for (unsigned int i = 0; i < stDeviceList.nDeviceNum; i++)
            {
                printf("[device %d]:\n", i);
                MV_CC_DEVICE_INFO* pDeviceInfo = stDeviceList.pDeviceInfo[i];
                if (NULL == pDeviceInfo)
                {
                    break;
                } 
                PrintDeviceInfo(pDeviceInfo);            
            }  
        } 
        else
        {
            printf("Find no devices!\n");
            break;
        }

        printf("Please Input camera index(0-%d):", stDeviceList.nDeviceNum-1);
        unsigned int nIndex = 0;
        scanf_s("%d", &nIndex);

        if (nIndex >= stDeviceList.nDeviceNum)
        {
            printf("Input error!\n");
            break;
        }

        // ch:选择设备并创建句柄 | en:Select device and create handle
        nRet = MV_CC_CreateHandle(&handle, stDeviceList.pDeviceInfo[nIndex]);
        if (MV_OK != nRet)
        {
            printf("Create handle fail! nRet [0x%x]\n", nRet);
            break;
        }

        // ch:打开设备 | en:Open device
        nRet = MV_CC_OpenDevice(handle);
        if (MV_OK != nRet)
        {
            printf("Open device fail! nRet [0x%x]\n", nRet);
            break;
        }

        // ch:探测网络最佳包大小(只对GigE相机有效) | en:Detection network optimal package size(It only works for the GigE camera)
        if (stDeviceList.pDeviceInfo[nIndex]->nTLayerType == MV_GIGE_DEVICE)
        {
            int nPacketSize = MV_CC_GetOptimalPacketSize(handle);
            if (nPacketSize > 0)
            {
                nRet = MV_CC_SetIntValueEx(handle,"GevSCPSPacketSize",nPacketSize);
                if(nRet != MV_OK)
                {
                    printf("Warning: set packet size fail nRet [0x%x]!\n", nRet);
                }
            }
            else
            {
                printf("Warning: Get packet size fail nRet [0x%x]!\n", nPacketSize);
            }
        }

        // 以下设置Line0 IO输入，也可选择Line2作为输入、Line2 延迟更低
        printf("Now set IO input...\n");
        {
            // ch:设置触发模式为on | en:Set trigger mode as on
            nRet = MV_CC_SetEnumValueByString(handle, "TriggerMode", "On");
            if (MV_OK != nRet)
            {
                printf("Set trigger mode fail! nRet [0x%x]\n", nRet);
                break;
            }

            // ch:设置触发源为Line0 | en:Set trigger source as Line0
            nRet = MV_CC_SetEnumValueByString(handle, "TriggerSource", "Line0");
            if (MV_OK != nRet)
            {
                printf("Set trigger source fail! nRet [0x%x]\n", nRet);
                break;
            }

            // ch:设置触发沿为RisingEdge | en: Set TriggerActivation as RisingEdge
            nRet = MV_CC_SetEnumValueByString(handle, "TriggerActivation", "RisingEdge");
            if (MV_OK != nRet)
            {
                printf("Set trigger activation RisingEdge fail! nRet [0x%x]\n", nRet);
                break;
            }

            // ch:设置触发延迟 | en: Set TriggerDelay
            nRet = MV_CC_SetFloatValue(handle, "TriggerDelay", 0);
            if (MV_OK != nRet)
            {
                printf("Set trigger delay fail! nRet [0x%x]\n", nRet);
                break;
            }

            // ch:关闭触发缓存、如需要可开启 | en: Turn off TriggerCacheEnable 
            nRet = MV_CC_SetBoolValue(handle, "TriggerCacheEnable", false);
            if (MV_OK != nRet)
            {
                printf("Set trigger cache enable fail! nRet [0x%x]\n", nRet);
                break;
            }

            // ch:切换LineSelector 为Line0 | en: Set LineSelector as Line0
            nRet = MV_CC_SetEnumValueByString(handle, "LineSelector", "Line0");
            if (MV_OK != nRet)
            {
                printf("Set line selector fail! nRet [0x%x]\n", nRet);
                break;
            }

            // ch:设置Line0 滤波时间(us)、误触发可适当加大 | Set Line0 LineDebouncerTime(us)
            nRet = MV_CC_SetIntValueEx(handle, "LineDebouncerTime", 50);
            if (MV_OK != nRet)
            {
                printf("Set line debouncer time fail! nRet [0x%x]\n", nRet);
                break;
            }
        }

        //以下设置Line1 IO输出、也可选择Line2作为输出、Line2 延迟更低；用于控制外部光源、等设备
        printf("Now set IO output...\n");
        {
            // ch:切换LineSelector 为Line1 | en:Set LineSelector as Line1
            nRet = MV_CC_SetEnumValueByString(handle, "LineSelector", "Line1");
            if (MV_OK != nRet)
            {
                printf("Set line selector fail! nRet [0x%x]\n", nRet);
                break;
            }
            //ch: 输出源选择曝光开始 |en:Set LineSource as ExposureStartActive
            nRet = MV_CC_SetEnumValueByString(handle, "LineSource", "ExposureStartActive");
            if (MV_OK != nRet)
            {
                printf("Set LineSource  fail! nRet [0x%x]\n", nRet);
                break;
            }
            // ch:开启输出使能| en: Turn on StrobeEnable
            nRet = MV_CC_SetBoolValue(handle, "StrobeEnable", true);
            if (MV_OK != nRet)
            {
                printf("Set strobe enable fail! nRet [0x%x]\n", nRet);
                break;
            }

            // ch:设置输出线路持续时间（us） | en: Set StrobeLineDuration(us)
            nRet = MV_CC_SetIntValueEx(handle, "StrobeLineDuration", 0);
            if (MV_OK != nRet)
            {
                printf("Set line strobe line duration fail! nRet [0x%x]\n", nRet);
                break;
            }
            // ch:设置输出线路延迟（us） | en: Set StrobeLineDelay(us)
            nRet = MV_CC_SetIntValueEx(handle, "StrobeLineDelay", 0);
            if (MV_OK != nRet)
            {
                printf("Set line strobe line delay fail! nRet [0x%x]\n", nRet);
                break;
            }
            // ch:输出线路预延迟（us） | en: Set StrobeLinePreDelay(us)
            nRet = MV_CC_SetIntValueEx(handle, "StrobeLinePreDelay", 0);
            if (MV_OK != nRet)
            {
                printf("Set line strobe line pre-delay fail! nRet [0x%x]\n", nRet);
                break;
            }
           
        }

		// ch:注册抓图回调 | en:Register image callback
		nRet = MV_CC_RegisterImageCallBackEx(handle, ImageCallBackEx, handle);
		if (MV_OK != nRet)
		{
			printf("Register Image CallBack fail! nRet [0x%x]\n", nRet);
			break;
		}

        // ch:开始取流 | en:Start grab image
        nRet = MV_CC_StartGrabbing(handle);
        if (MV_OK != nRet)
        {
            printf("Start grabbing fail! nRet [0x%x]\n", nRet);
            break;
        }
        else
        {
            printf("Start grabbing success!\n");
        }

        printf("Press a key to stop.\n");
        WaitForKeyPress();

        // ch:停止取流 | en:Stop grab image
        nRet = MV_CC_StopGrabbing(handle);
        if (MV_OK != nRet)
        {
            printf("Stop grabbing fail! nRet [0x%x]\n", nRet);
            break;
        }

		// ch:注销抓图回调 | en:Unregister image callback
		nRet = MV_CC_RegisterImageCallBackEx(handle, NULL, NULL);
		if (MV_OK != nRet)
		{
			printf("Unregister image callBack fail! nRet [0x%x]\n", nRet);
			break;
		}

        // ch:关闭设备 | Close device
        nRet = MV_CC_CloseDevice(handle);
        if (MV_OK != nRet)
        {
            printf("Close device fail! nRet [0x%x]\n", nRet);
            break;
        }

        // ch:销毁句柄 | Destroy handle
        nRet = MV_CC_DestroyHandle(handle);
        if (MV_OK != nRet)
        {
            printf("Destroy handle fail! nRet [0x%x]\n", nRet);
            break;
        }
		handle = NULL;
    } while (0);
 
	if (handle != NULL)
	{
		MV_CC_DestroyHandle(handle);
		handle = NULL;
	}
    

	// ch:反初始化SDK | en:Finalize SDK
	MV_CC_Finalize();

    printf("Press a key to exit.\n");
    WaitForKeyPress();

    return 0;
}
